addons/sale_loyalty/tests/test_program_rules.py

```diff
@@ -231,36 +229,86 @@ def test_program_rules_validity_dates_and_uses(self):
             })
         ]})
         self._auto_rewards(order, self.immediate_promotion_program)
-        self.assertEqual(len(order.order_line.ids), 2, "The promo offer shouldn't have been applied we're not between the validity dates")
+        self.assertEqual(len(order.order_line.ids), 2,
+                         "The promo shouldn't have been applied as it is expired.")
 
         self.immediate_promotion_program.write({
             'date_to': date.today() + timedelta(days=2),
         })
-        order = self.env['sale.order'].create({'partner_id': self.steve.id})
+
+        self._auto_rewards(order, self.immediate_promotion_program)
+        self.assertEqual(len(order.order_line.ids), 3,
+                         "The promo should have been applied we're between the validity dates.")
+
+        # Test date_from (no date_to)
+        self.immediate_promotion_program.write({
+            'date_from': date.today() + timedelta(days=2),
+            'date_to': False,
+        })
+
+        self._auto_rewards(order, self.immediate_promotion_program)
+        self.assertEqual(len(order.order_line.ids), 2,
+                         "The promo shouldn't have been applied as it is not active yet.")
+
+        self.immediate_promotion_program.write({
+            'date_from': date.today() - timedelta(days=2),
+        })
+
+        self._auto_rewards(order, self.immediate_promotion_program)
+        self.assertEqual(len(order.order_line.ids), 3,
+                         "The promo should have been applied we're between the validity dates.")
```

Comment:
```suggestion
        self.immediate_promotion_program.write({
            'date_from': date.today() - timedelta(days=2),
        })
        self._auto_rewards(order, self.immediate_promotion_program)
        self.assertEqual(len(order.order_line.ids), 3,
                         "The promo should have been applied we're between the validity dates.")
```

Let's make a single block for every case, it's easier to read =]